File: report/src/main/java/com/vmlens/report/runelementtype/memoryaccesskey/ArrayObjectHashCodeAndIndex.java
package com.vmlens.report.runelementtype.memoryaccesskey;

import com.vmlens.report.description.DescriptionContext;
import com.vmlens.report.description.NeedsDescriptionCallback;

public class ArrayObjectHashCodeAndIndex implements MemoryAccessKey  {

    private final long objectHashCode;
    private final int arrayIndex;

    public ArrayObjectHashCodeAndIndex(long objectHashCode, int arrayIndex) {
        this.objectHashCode = objectHashCode;
        this.arrayIndex = arrayIndex;
    }

    @Override
    public String asString(DescriptionContext context) {
        return " array[" + arrayIndex  + "]@" + objectHashCode;
    }

    @Override
    public void addToNeedsDescription(NeedsDescriptionCallback callback) {
   
    }
    
}

File: sync-bug/src/main/scala/com/anarsoft/race/detection/report/element/runelementtype/memoryaccesskey/ArrayObjectHashCodeAndIndex.scala
package com.anarsoft.race.detection.report.element.runelementtype.memoryaccesskey

import com.anarsoft.race.detection.report.description.{DescriptionContext, NeedsDescriptionCallback};

class ArrayObjectHashCodeAndIndex( val objectHashCode: Long, val arrayIndex: Int) extends MemoryAccessKey {
  
  override def asString(context: DescriptionContext): String = " array[" + arrayIndex + "]"

  override def addToNeedsDescription(callback: NeedsDescriptionCallback): Unit = {
  }

  override def objectHashCodeOption: Option[Long] = Some(objectHashCode);


  private def canEqual(other: Any): Boolean = other.isInstanceOf[ArrayObjectHashCodeAndIndex]
  
  override def equals(other: Any): Boolean = other match {
    case that: ArrayObjectHashCodeAndIndex =>
      that.canEqual(this) &&
        objectHashCode == that.objectHashCode &&
        arrayIndex == that.arrayIndex
    case _ => false
  }
  
  override def hashCode(): Int = {
    val state = Seq(objectHashCode, arrayIndex)
    state.map(_.hashCode()).foldLeft(0)((a, b) => 31 * a + b)
  }
}
